docs/development/pigs/pig-009.rst

```diff
@@ -0,0 +1,298 @@
+multiple PDFs at the same time to avoid a Python loop over the PDFs. This will 
+be supported by specifying the `axis` argument:
+
+.. code::
+
+    cdf_sampler = InverseCDFSampler(pdf=pdfs, axis=0)
+    samples = cdf_sampler.sample(size=1)
+
+
+The `InverseCDFSampler` could addtionally implement a few helper functions to plot
+the CDF and PDF. The `InverseCDFSampler` class will replace the current
+`GeneralRandom` and `GeneralRandomArray` classes in `gammapy.utils.distributions`. 
+
+
+
+MapEventSampler
+---------------
+
+To support sampling from the predicted number of counts maps for source as well 
+as background maps, we propose to introduce a `MapEventSampler` class. The 
+responsibility of this class is to handle the physical coordinate transformations, 
+time dependence of the model, output data structures and sampling of the total 
+number of observed events.
+Starting from a predicted number of counts map (e.g. computed by the `MapEvaluator`):
+
+.. code::
+
+    src = SkyModel()
+    evaluator = MapEvaluator(src, exposure, psf=None, edisp=None)
+    npred_map = evaluator.compute_npred()
+
+The `MapEventSampler` takes this map and a lightcurve model as an input and 
+allows to sample from it with the `.sample()` method:
+
+.. code::
+
+    lightcurve = LightCurveTableModel() # or PhaseCurveTableModel()
+    src_sampler = MapEventSampler(npred_map, lightcurve, t_min=, t_max=)
+
+    # draw total number of events
+    n_events = src_sampler.npred_total()
+
+    # draw events as an astropy table
+    events_src = map_sampler.sample()
+
+If no lightcurve is provided a constant rate can be assumed e.g. for 
+background models:
+
+.. code::
+
+    bkg_map = BackgroundModel().map
+    bkg_sampler = MapEventSampler(bkg_map, t_min=, t_max=)
+    events_bkg = bkg_sample.sample()
+
+
+The `MapEventSampler` returns an `astropy.table.Table` object, with the following
+columns: `TIME`, `RA_TRUE`, `DEC_TRUE`, `ENERGY_TRUE`. Later additional info such
+as `DETX` and `DETY` could be added.
+
+
+IRFEventDistributor
+-------------------
+
+We also propose a class to handle correctly the spatially varying IRFs: the
+`IRFEventDistibutor` class. Based on the `PSFMap` and `EDispMap`
+as an input it interpolates the "correct" IRF at the position of a given event and
+applies it. In more detail the class will calculate the psf and the energy dispersion 
+for each event true position and true energy. The IRFs are assumed to be constant
+and not time-dependent. The time dependency will be handled higher level, by a loop
+over good time-intervals (GTIs). The class would be used as following:
+
+.. code::
+
+    events = vstack([events_src_1, events_src_2, ..., events_src_N])
+    event_distributor = IRFEventDistributor(psf_map, edisp_map)
+
+    events = irf_distributor.sample_psf(events)
+    events = irf_distributor.sample_edisp(events)
+
+Alternatively the IRFs could be applied per model component and be handled in 
```

Comment:
Alternatives should go in the Alternatives / Outlook section at the end.
Try to keep the proposal short, really describing what you plan to implement in the first version you build.
Like anything in Gammapy, this event sampling code will evolve and improve over the coming months and years.
But for the PIG to be really useful, we have to focus discussion on what to build now as a first version, for v1.0.